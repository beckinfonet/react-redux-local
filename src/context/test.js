import 'raf/polyfill'
import React from 'react'
import prettyFormat from 'pretty-format'
import { render, fireEvent } from 'react-testing-library'

import { createContext } from '../..'

import { reducer, actions } from '../duck'
import { combinations } from '../utils'

it('should be defined', () => {
  expect(createContext).toBeDefined()
})

const createWrapper = ({ mapState, mapActions }) => {
  const { Provider, Consumer } = createContext({
    reducer,
    actions
  })

  return render(
    <Provider>
      <Consumer mapState={mapState} mapActions={mapActions}>
        {(state, actions = {}) => {
          const inc = actions.increase || function() {}
          const dec = actions.decrease || function() {}
          const reset = actions.reset || function() {}

          return (
            <div>
              Given:
              {prettyFormat([
                mapState ? mapState(reducer()) : mapState,
                mapActions ? mapActions(actions) : mapActions
              ])}
              Got:
              {prettyFormat([state, actions])}
              <button onClick={inc} type="button">
                inc
              </button>
              <button onClick={dec} type="button">
                dec
              </button>
              <button onClick={reset} type="button">
                reset
              </button>
            </div>
          )
        }}
      </Consumer>
    </Provider>
  )
}

const tests = combinations({
  mapState: [undefined, state => state, state => state.counter],
  mapActions: [
    undefined,
    actions => actions,
    actions => ({ increase: actions.increase })
  ]
})

describe('should render with different selectors', () => {
  tests.map(props =>
    test('[variation]', () => {
      const { container } = createWrapper(props)
      expect(container).toMatchSnapshot()
    })
  )
})

test('should be able to update state by firing an action', () => {
  const props = {
    mapState: state => state,
    mapActions: actions => actions
  }

  const { container, getByText } = createWrapper(props)
  fireEvent.click(getByText('inc'))
  expect(container).toMatchSnapshot()
  fireEvent.click(getByText('reset'))
  expect(container).toMatchSnapshot()
  fireEvent.click(getByText('dec'))
  expect(container).toMatchSnapshot()
})
