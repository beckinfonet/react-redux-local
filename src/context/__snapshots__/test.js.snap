// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to update state by firing an action 1`] = `
<div>
  <div>
    Given:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    Got:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should be able to update state by firing an action 2`] = `
<div>
  <div>
    Given:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    Got:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should be able to update state by firing an action 3`] = `
<div>
  <div>
    Given:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    Got:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 1`] = `
<div>
  <div>
    Given:
    Array [
  undefined,
  undefined,
]
    Got:
    Array [
  undefined,
  Object {},
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 2`] = `
<div>
  <div>
    Given:
    Array [
  undefined,
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    Got:
    Array [
  undefined,
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 3`] = `
<div>
  <div>
    Given:
    Array [
  undefined,
  Object {
    "increase": [Function anonymous],
  },
]
    Got:
    Array [
  undefined,
  Object {
    "increase": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 4`] = `
<div>
  <div>
    Given:
    Array [
  Object {
    "counter": 0,
  },
  undefined,
]
    Got:
    Array [
  Object {
    "counter": 0,
  },
  Object {},
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 5`] = `
<div>
  <div>
    Given:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    Got:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 6`] = `
<div>
  <div>
    Given:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "increase": [Function anonymous],
  },
]
    Got:
    Array [
  Object {
    "counter": 0,
  },
  Object {
    "increase": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 7`] = `
<div>
  <div>
    Given:
    Array [
  0,
  undefined,
]
    Got:
    Array [
  0,
  Object {},
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 8`] = `
<div>
  <div>
    Given:
    Array [
  0,
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    Got:
    Array [
  0,
  Object {
    "decrease": [Function anonymous],
    "increase": [Function anonymous],
    "reset": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;

exports[`should render with different selectors [variation] 9`] = `
<div>
  <div>
    Given:
    Array [
  0,
  Object {
    "increase": [Function anonymous],
  },
]
    Got:
    Array [
  0,
  Object {
    "increase": [Function anonymous],
  },
]
    <button
      type="button"
    >
      inc
    </button>
    <button
      type="button"
    >
      dec
    </button>
    <button
      type="button"
    >
      reset
    </button>
  </div>
</div>
`;
